File: 백준/Silver/3085. 사탕 게임/사탕 게임.py
n = int(input())
board = [list(input()) for _ in range(n)]

dx = [1, 0]
dy = [0, -1]

def check():
    max_count = 0
    for i in range(n):
        cnt1 = 1
        for j in range(1, n):
            if board[i][j] == board[i][j-1]:
                cnt1 += 1
            else:
                cnt1 = 1
            max_count = max(cnt1, max_count)
        cnt2 = 1
        for j in range(1, n):
            if board[j][i] == board[j-1][i]:
                cnt2 += 1
            else:
                cnt2 = 1
            max_count = max(cnt2, max_count)
                
    return max_count

result = 0
for i in range(n):
    for j in range(n):
        for k in range(2):
            nx = i + dx[k]
            ny = j + dy[k]
            if 0 <= nx < n and 0 <= ny < n and board[nx][ny] != board[i][j]:
                board[i][j], board[nx][ny] = board[nx][ny], board[i][j]
                result = max(result, check())
                board[i][j], board[nx][ny] = board[nx][ny], board[i][j]
print(result)

File: 백준/Silver/3085. 사탕 게임/사탕 게임.java
import java.util.Arrays;
import java.util.Scanner;


public class Main {
    public static int solution(int n, char[][] arr) {
        int[] dx = {1, 0};
        int[] dy = {0, 1};

        int result = 0;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                for (int k = 0; k < 2; k++) {
                    int nx = i + dx[k];
                    int ny = j + dy[k];
                    if (nx >= 0 && nx < n && ny >= 0 && ny < n && arr[nx][ny] != arr[i][j]) {
                        char tmp = arr[nx][ny];
                        arr[nx][ny] = arr[i][j];
                        arr[i][j] = tmp;
                        result = Math.max(result, check(n, arr));
                        tmp = arr[nx][ny];
                        arr[nx][ny] = arr[i][j];
                        arr[i][j] = tmp;
                    }
                }
            }
        }
        return  result;
    }

    public static int check(int n, char[][] arr) {
        int result = 0;
        for (int i = 0; i < n; i++) {
            int sum = 1;
            for (int j = 0; j < n; j++) {
                if (j + 1 < n && arr[i][j] == arr[i][j + 1]) {
                    sum += 1;
                } else {
                    result = Math.max(sum, result);
                    sum = 1;
                }
            }
            result = Math.max(sum, result);
        }

        for (int i = 0; i < n; i++) {
            int sum = 1;
            for (int j = 0; j < n; j++) {
                if (j + 1 < n && arr[j][i] == arr[j+ 1][i]) {
                    sum += 1;
                } else {
                    result = Math.max(sum, result);
                    sum = 1;
                }
            }
            result = Math.max(sum, result);
        }
        return result;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        int n = scanner.nextInt();

        char[][] array = new char[n][n];
        for (int i = 0; i < n; i++) {
            char[] arr = scanner.next().toCharArray();
            for (int j = 0; j< n; j++) {
                array[i][j] = arr[j];
            }
        }
        System.out.println(solution(n, array));

    }

}